Offline/OfflinePlayer/objc/OfflinePlayer/VideosViewController.m

```diff
@@ -0,0 +1,1370 @@
+    if (offlineVideoPath == nil)
+    {
+        return @"MediaSelection(n/a)";
+    }
+
+    NSURL *offlineVideoPathURL = [NSURL fileURLWithPath:offlineVideoPath];
+    
+    AVURLAsset *URLAsset = [AVURLAsset assetWithURL:offlineVideoPathURL];
+    
+    NSString *description = [self mediaSelectionDescription:mediaSelection
+                                                   URLAsset:URLAsset];
+    return description;
+}
+
+- (NSString *)mediaSelectionDescription:(AVMediaSelection *)mediaSelection
+                               URLAsset:(AVURLAsset *)URLAsset
+{
+    // Return a string description of the specified Media Selection.
+    AVMediaSelectionGroup *legibleMediaSelectionGroup = [URLAsset mediaSelectionGroupForMediaCharacteristic:AVMediaCharacteristicLegible];
+    AVMediaSelectionGroup *audibleMediaSelectionGroup = [URLAsset mediaSelectionGroupForMediaCharacteristic:AVMediaCharacteristicAudible];
+    AVMediaSelectionOption *legibleMediaSelectionOption = [mediaSelection selectedMediaOptionInMediaSelectionGroup:legibleMediaSelectionGroup];
+    AVMediaSelectionOption *audibleMediaSelectionOption = [mediaSelection selectedMediaOptionInMediaSelectionGroup:audibleMediaSelectionGroup];
+    
+    NSString *description = [NSString stringWithFormat:@"MediaSelection(obj:%p, legible:%@, audible:%@)",
+                             mediaSelection,
+                             legibleMediaSelectionOption.displayName ?: @"-",
+                             audibleMediaSelectionOption.displayName ?: @"-"];
+    return description;
+}
+
+- (void)createPlayerView
+{
+    // The player view is the BrightCove PlayerUI with built-in controls.
+    // This is where the video will be presented; it's reused for all videos.
+    if (self.playerView == nil)
+    {
+        BCOVPUIPlayerViewOptions *options = [[BCOVPUIPlayerViewOptions alloc] init];
+        options.presentingViewController = self;
+        
+        BCOVPUIBasicControlView *controlView = [BCOVPUIBasicControlView basicControlViewWithVODLayout];
+        self.playerView = [[BCOVPUIPlayerView alloc] initWithPlaybackController:nil
+                                                                        options:options
+                                                                   controlsView:controlView ];
+
+        self.playerView.frame = self.videoContainer.bounds;
+        self.playerView.autoresizingMask = (UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth);
+        self.playerView.delegate = self;
+        [self.videoContainer addSubview:self.playerView];
+    }
+}
+
+- (void)createNewPlaybackController
+{
+    if (!self.playbackController)
+    {
+        NSLog(@"Creating a new playbackController");
+
+        BCOVPlayerSDKManager *sdkManager = [BCOVPlayerSDKManager sharedManagerWithOptions:@{@"log_level": @1 }]; // level one shows logs errors to the debug console
+
+        // Publisher/application IDs not required for Dynamic Delivery
```

Comment:
Insert comment:
// This app shows how to set up your playback controller for playback of FairPlay-protected videos.
// This playback controller, as well as the download manager will work with either FairPlay-protected
// videos, or "clear" videos (no DRM protection).

Reply:
√